src/core/experimental/array.d

```diff
@@ -0,0 +1,2175 @@
+    }
+
+    private void destroyUnused()
+    {
+        if (support !is null)
+        {
+            delRef(support);
+        }
+    }
+
+    /**
+     * Constructs a qualified array out of a number of items
+     * that will use the collection deciced allocator object.
+     *
+     * Params:
+     *      values = a variable number of items, either in the form of a
+     *               list or as a built-in array
+     *
+     * Complexity: $(BIGOH m), where `m` is the number of items.
+     */
+    this(U, this Q)(U[] values...)
+    if (!is(Q == shared) && is(U : T))
+    {
+        static if (is(Q == immutable) || is(Q == const))
+        {
+            static if (is(Q == immutable))
+            {
+                bool _isShared = true;
+                mixin(immutableInsert!(typeof(values), true, "values")());
+                setIsShared(support, true);
+            }
+            else
+            {
+                bool _isShared = false;
+                mixin(immutableInsert!(typeof(values), true, "values")());
+            }
+        }
+        else
+        {
+            insert(0, values);
+        }
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        // Create a list from a list of ints
+        {
+            auto a = rcarray!int(1, 2, 3);
+            assert(a == [1, 2, 3]);
+        }
+        // Create a list from an array of ints
+        {
+            auto a = rcarray!int([1, 2, 3]);
+            assert(a == [1, 2, 3]);
+        }
+        // Create a list from a list from an input range
+        {
+            auto a = rcarray!int(1, 2, 3);
+            auto a2 = rcarray!int(a);
+            assert(a2 == [1, 2, 3]);
+        }
+    }
+
+    // Begin Copy Ctors
+    // {
+
+    private enum copyCtorIncRef = q{
+        payload = rhs.payload;
+        support = rhs.support;
+        //isShared = rhs.isShared;
+
+        if (support !is null)
+        {
+            setIsShared(support, rhs.isShared);
+            addRef(support);
+        }
+    };
+
+    this(ref typeof(this) rhs)
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    // { Get a const obj
+
+    this(ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(const ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(immutable ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+    // } Get a const obj
+
+    // { Get an immutable obj
+
+    this(ref typeof(this) rhs) immutable
+    {
+        bool _isShared = true;
+        mixin(immutableInsert!(typeof(rhs), true, "rhs")());
+        //isShared = true;
+        setIsShared(support, true);
+    }
+
+    this(const ref typeof(this) rhs) immutable
+    {
+        bool _isShared = true;
+        mixin(immutableInsert!(typeof(rhs.payload), true, "rhs.payload")());
+        //isShared = true;
+        setIsShared(support, true);
+    }
+
+    this(immutable ref typeof(this) rhs) immutable
```

Comment:
Changing to
`    this(scope immutable ref typeof(this) rhs) immutable` fix the `scope` error, but will make the ctor `@system`.
Adding a `@trusted` (For testing purposes only) to quickly mute all the complaints will pass all the tests. Why is the `scope` required and why is it making the function `@system` when used?

Reply:
Another way of bypassing the error, without any `@trusted` "magic" is to change the `idup` implementation from
```
immutable(rcarray!T) idup(this Q)()
{
    return immutable rcarray!T(this);
}
```
to
```
immutable(rcarray!T) idup(this Q)()
{
    static if (is(Q == immutable))
    {
        return this;
    }
    else
    {
        return immutable rcarray!T(this);
    }
}
```

Reply:
The generated copy constructor has the signature "this(ref scope return inout typeof(this) ths) inout". The error appears if the rcarray is a member of a struct for which a copy constructor would be implicitly defined: I assume that the generated copy constructor calls your defined copy constructor.